### PR TITLE
Enable MDSPAN support by default

### PR DESCRIPTION
**Pull request by crtrott**

### CODE REVIEW

**Comment by crtrott** [comment]
Retest this please!

**Review by nliber — APPROVED** [review]

**Comment by ndellingwood** [comment]
@crtrott it looks like this will require explicitly disabling mdspan in Trilinos builds? I'm guessing this will be necessary to support e.g. Sacado and Stokhos

**Comment by crtrott** [comment]
@ndellingwood no for now this does NOT replace internals of Kokkos::View it just enables us testing mdspan interop like the conversion operators.

**Review by masterleinad — APPROVED** [review]

**Comment by crtrott** [comment]
We also likely need to update mdspan TPL first for some windows fixes

**Review by masterleinad — CHANGES_REQUESTED** [review]
A bunch of builds are failing with
```
    run_test_mirror();

    ^

/var/jenkins/workspace/Kokkos_PR-7069/containers/unit_tests/TestDynViewAPI_generic.hpp:20:43: note: in instantiation of member function 'Test::TestDynViewAPI<double, Kokkos::Experimental::SYCL>::run_tests' requested here
  TestDynViewAPI<double, TEST_EXECSPACE>::run_tests();
                                          ^
1 error generated.
    run_test_mirror();
    ^
/var/jenkins/workspace/Kokkos_PR-7069/containers/unit_tests/TestDynViewAPI_generic.hpp:20:43: note: in instantiation of member function 'Test::TestDynViewAPI<double, Kokkos::Experimental::SYCL>::run_tests' requested here
  TestDynViewAPI<double, TEST_EXECSPACE>::run_tests();
                                          ^
1 error generated.
```
Looks like you need to update the bundled mdspan version.

**Comment by crtrott** [comment]
ok i reenabled in Windows and updated mdspan to current stable

**Comment by masterleinad** [comment]
> ok i reenabled in Windows and updated mdspan to current stable

Looks like we need https://github.com/kokkos/mdspan/commit/421d2c209bf669d58fe8d1b3d035203a6d8d4176 for MSVC.

**Comment by masterleinad** [comment]
`Visual Studio 16 2019` seems to have more problems like
```
C:\projects\source\tpls\mdspan\include\mdspan\../experimental/__p2642_bits/layout_padded.hpp(76,15): error C2672: 'Kokkos::Experimental::detail::get_actual_static_padding_value': no matching overloaded function found [C:\projects\source\build\core\src\kokkoscore.vcxproj]
C:\projects\source\tpls\mdspan\include\mdspan\../experimental/__p2642_bits/layout_padded.hpp(78): message : see reference to class template instantiation 'Kokkos::Experimental::detail::static_array_type_for_padded_extent<_PaddingValue,_Extents,_ExtentToPadIdx,_Rank,Enabled>' being compiled [C:\projects\source\build\core\src\kokkoscore.vcxproj]
C:\projects\source\tpls\mdspan\include\mdspan\../experimental/__p2642_bits/layout_padded.hpp(77,1): error C2975: '_PaddingValue': invalid template argument for 'Kokkos::Experimental::detail::get_actual_static_padding_value', expected compile-time constant expression [C:\projects\source\build\core\src\kokkoscore.vcxproj]
C:\projects\source\tpls\mdspan\include\mdspan\../experimental/__p2642_bits/layout_padded.hpp(43): message : see declaration of '_PaddingValue' [C:\projects\source\build\core\src\kokkoscore.vcxproj]
  Kokkos_Error.cpp
```

**Comment by crtrott** [comment]
Yeah I just confirmed it works all with VS2022 but not 2019 :-(

**Comment by masterleinad** [comment]
We still need https://github.com/kokkos/mdspan/commit/421d2c209bf669d58fe8d1b3d035203a6d8d4176 for the Windows 2022 MSVC+Cuda build.

**Review by masterleinad — APPROVED** [review]

**Review by nmm0 — APPROVED** [review]
Looks good

File: core/unit_test/TestMDSpanConversion.hpp [hunk 1]
```diff
@@ -120,12 +120,15 @@ struct TestViewMDSpanConversion {
...
+          typename ViewType::memory_space,
+          Kokkos::default_accessor<
+              const typename natural_mdspan_type::element_type>>;
       using mdspan_type = Kokkos::mdspan<
           const typename natural_mdspan_type::element_type,
           Kokkos::dextents<typename natural_mdspan_type::index_type,
                            natural_mdspan_type::rank()>,
-          typename natural_mdspan_type::layout_type,
-          typename natural_mdspan_type::accessor_type>;
+          typename natural_mdspan_type::layout_type, const_acc_type>;
```
Comment (nmm0) [inline core/unit_test/TestMDSpanConversion.hpp:131]: Isn't this the same bug fixed by https://github.com/kokkos/kokkos/pull/7070

We may want to have a more general solution if we add other accessor types

Comment (crtrott) [reply core/unit_test/TestMDSpanConversion.hpp]: yeah I rebased now to have this all sorted out

**Comment by ndellingwood** [comment]
Should this get a changelog entry for 4.4?

**Comment by dalg24** [comment]
> Should this get a changelog entry for 4.4?

I'd say we can refer to it from the entry mentioning conversion to and from mdspan.  I will do it.

**Pull request merged**